### PR TITLE
Implement highest and lowest function families

### PR DESCRIPTION
**Pull request by stivenbb**
Native implementation of the following Graphite functions:
[highest()](http://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.highest)
[highestAverage()](http://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.highestAverage)
[highestCurrent()](http://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.highestCurrent)
[highestMax()](http://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.highestMax)
[lowest()](http://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.lowest)
[lowestAverage()](http://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.lowestAverage)
[lowestCurrent()](http://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.lowestCurrent)

### CODE REVIEW

**Comment by stivenbb** [comment]
Here is a speed improvement report:

```
===== Running Speed Test on highestlowest.json =====
---------- Native highestlowest.json Latencies ----------
Mean: 214.763444ms
50th percentile: 78.671643ms
95th percentile: 890.445966ms
99th percentile: 1.710930491s
Max: 6.078288596s
Success: 100%
Errors: []

---------- Graphite (Python) highestlowest.json Latencies ----------
Mean: 2.273897245s
50th percentile: 477.845065ms
95th percentile: 11.34453194s
99th percentile: 14.43197089s
Max: 23.345060363s

Success: 100%
Errors: []

---------- Speed Improvement ----------
Mean: x10
50th percentile: x6
95th percentile: x12
99th percentile: x8
Max: x3
```

File: expr/func_highestlowest.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+			ArgSeriesList{val: &s.in},
+			ArgInt{key: "n", val: &s.n},
+		}, []Arg{ArgSeriesList{}}
+	}
+	return []Arg{
+		ArgSeriesList{val: &s.in},
+		ArgInt{key: "n", val: &s.n},
+		ArgString{key: "func", val: &s.fn, validator: []Validator{IsConsolFunc}},
+	}, []Arg{ArgSeriesList{}}
+}
```
Comment (Dieterbe) [inline expr/func_highestlowest.go]: this is clever. supporting different signatures.

File: expr/func_highestlowest.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+	if len(series) == 0 {
+		return series, nil
+	}
+
+	consolidationFunc := consolidation.GetAggFunc(consolidation.FromConsolidateBy(s.fn))
+
+	// Calculates the consolidated value for each series and stores it in a map
+	// where the key is a pointer to the first datapoint
+	consolidationVals := make(map[*schema.Point]float64, len(series))
```
Comment (Dieterbe) [inline expr/func_highestlowest.go]: why pointers? i would just use the float64 as map key

Comment (stivenbb) [reply expr/func_highestlowest.go]: The pointer is to the first Datapoint of the series. (not sure what float64 you're referring to)

Comment (Dieterbe) [reply expr/func_highestlowest.go]: the float64 of the first Datapoint of the series

Comment (stivenbb) [reply expr/func_highestlowest.go]: I'm not sure how that would work.. two series can have equal values in the first datapoint, so it wouldn't work as a map key.

**Comment by Dieterbe** [comment]
the scoring stuff was a bit too confusing, so i made it a tad more verbose, but more simple/standard.
see the commit i made at https://github.com/grafana/metrictank/commits/highestlowest does this look good to you? if so i'll merge.

**Comment by Dieterbe** [comment]
i applied your suggestion @shanson7 and merged it. thanks @stivenbb !

**Pull request merged**